config/decoder.go

```diff
@@ -205,17 +212,21 @@ func (d *decoder) mapping(childKey string, value reflect.Value, def string) erro
 	// child yamlNode parsed from yaml file is of type map[interface{}]interface{}
 	// type casting here makes sure that we are iterating over a parsed map.
 	if v, ok := val.(map[interface{}]interface{}); ok {
-		childKey += _separator
+		childKey = addSeparator(childKey)
+
 		for key := range v {
-			mapKey := childKey + fmt.Sprintf("%v", key)
-			itemValue := reflect.New(valueType.Elem()).Elem()
 
-			// Try to unmarshal value and save it in the map.
-			if err := d.unmarshal(mapKey, itemValue, def); err != nil {
-				return err
-			}
+			// Skip empty keys, because lookups in provider can toggle infinite loops/spill internals.
+			if mapKey := childKey + fmt.Sprintf("%v", key); mapKey != "" {
```

Comment:
suggest adding the empty check here and in addSeparator into a new PR so we can test it out.